### PR TITLE
Add equality field IDs to DeleteFile

### PR DESCRIPTION
**Pull request by rdblue**
This adds a list of equality field IDs to `DeleteFile` metadata that is tracked in manifests.

The equality field ID list identifies the subset of a delete file's fields (by ID) that should be used when comparing rows against the rows deleted in an equality delete file. The remaining fields are informational and store the values that were deleted by an equality delete. These may be used, for example, to reconstruct CDC events that contained deleted row values.

Additional fields will also be used for min/max stats filtering and are required to be accurate. When finding files for a table scan, the min/max values for non-comparison columns can still be used to filter delete files, even if the delete file applies to a data file that will be read.

For example, if a table contains columns `id` and `data`, a delete file may delete the row `(5, "bees!")` using only the `id` field for equality. A scan filter like `data like 'a%'` can cause the delete file to be ignored, even for a data file that contains ID `5` because the scan filter will remove the row.

### CODE REVIEW

**Comment by rdblue** [comment]
@rymurr, @prodeezy, @aokolnychyi, here's another one for row-level deletes. After this one and #1316 are in, we should be able to get end-to-end tests working.

**Comment by rdblue** [comment]
@openinx, @JingsongLi, @chenjunjiedada, the purpose of this is to support the delete file writers in #1316. Those writers support writing delete files using original data rows so that the deleted rows are available for streaming uses. Please have a look at both.

File: core/src/main/java/org/apache/iceberg/BaseFile.java [hunk 1]
```diff
@@ -385,7 +399,24 @@ public String toString() {
         .add("lower_bounds", lowerBounds)
         .add("upper_bounds", upperBounds)
         .add("key_metadata", keyMetadata == null ? "null" : "(redacted)")
-        .add("split_offsets", splitOffsets == null ? "null" : splitOffsets)
+        .add("split_offsets", splitOffsets == null ? "null" : splitOffsets())
+        .add("equality_ids", equalityIds == null ? "null" : equalityFieldIds())
         .toString();
   }
+
+  private static int[] toIntArray(List<Integer> ints) {
```
Comment (JingsongLi) [inline core/src/main/java/org/apache/iceberg/BaseFile.java:407]: If the data class pattern in Iceberg is:
- stores (primitive) arrays in class field.
- provides List method to return this field.

Can we add `ArrayUtil` class to provide `toPrimitive` and `toList` utils?

**Comment by JingsongLi** [comment]
Hi @rdblue , thanks for your work, these two PRs look very good~

I have two comments:

## Optimization for Upsert data

Considering that upsert data will write insert file and delete file at the same time,  this can double the storage.

I'm thinking about some scenarios:
-For example, the downstream does not need to restore CDC data stream.
-For example, downstream engines only need PKs(equality field IDs) for delete records.

How can we reduce storage in these scenarios? Can these additional fields be nulls?

## Why equality field IDs in `DeleteFile`?

Why not just primary keys definition for table? Will equality field IDs be different between files? 
It can be used as schema evolution?

**Comment by aokolnychyi** [comment]
> How can we reduce storage in these scenarios? Can these additional fields be nulls?

I had the same question. I think they should be null in most cases to reduce the size of delete files. It is good to have a flexible format, though.

> Why not just primary keys definition for table? Will equality field IDs be different between files?
It can be used as schema evolution?

This seems like an unnecessary restriction. Users would need to know all the columns they will update by at the table creation time and include them into the composite natural key. While it is essential to sort the data by the upsert key for performance, we can still append an extra column later on.

**Comment by aokolnychyi** [comment]
Apart from comments by @JingsongLi and @chenjunjiedada, this PR looks good to me.

**Comment by rdblue** [comment]
> Can these additional fields be nulls?

Yes. In fact, the fields can be omitted from the delete file entirely. There is no requirement to have any fields other than the ones in the delete ID set. We may also want to build a way to pass a row to get stats for the delete file, but not actually store the deleted values. But I think optimizations like this can be done in follow-up PRs.

> Why equality field IDs in DeleteFile?

When you encode a delete, the equality columns are fixed. As Anton noted, the columns you use may change as the table schema evolves. In those cases, we need to make sure that each delete uses equality columns from when it was committed. Storing them in `DeleteFile` metadata is a good way to do that.

In addition to being able to change the column set for schema evolution, this supports being able to delete by different columns if needed. For example, a GDPR deletes may come in by email address or by user name and we want to support encoding both without scanning to find out the file/position to delete.

Lastly, we could store these just in file metadata. But keeping them at the table level enables a few things. First, we can group files that use the same delete fields into the same filter set. Second, we can use them at job planning time to check whether a data file and a delete file overlap using delete column stats. But we can only do that if we know which columns are used to delete at job planning time.

File: api/src/main/java/org/apache/iceberg/ContentFile.java [hunk 2]
```diff
@@ -97,6 +97,17 @@
    */
   List<Long> splitOffsets();
 
+  /**
+   * Returns the set of field IDs used for equality comparison, in equality delete files.
+   * <p>
+   * An equality delete file may contain additional data fields that are not used by equality
+   * comparison. The subset of columns in a delete file to be used in equality comparison are
+   * tracked by ID. Extra columns can be used to
```
[the remark this answers is not in the document]
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/ContentFile.java]: Fixed. Thanks!

File: core/src/main/java/org/apache/iceberg/BaseFile.java (hunk 1 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/BaseFile.java]: Moved to `ArrayUtil`.

File: api/src/main/java/org/apache/iceberg/ContentFile.java [hunk 3]
```diff
@@ -102,6 +102,18 @@
    */
   List<Long> splitOffsets();
 
+  /**
+   * Returns the set of field IDs used for equality comparison, in equality delete files.
+   * <p>
+   * An equality delete file may contain additional data fields that are not used by equality
+   * comparison. The subset of columns in a delete file to be used in equality comparison are
+   * tracked by ID. Extra columns can be used to reconstruct changes and metrics from extra
```
Comment (rdsr) [inline api/src/main/java/org/apache/iceberg/ContentFile.java:110]: I think there might be a minor typo here - maybe after 'from extra'

Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/ContentFile.java:110]: I think it was intended. The second half of the sentence is "Metrics from extra columns are used during job planning". We might be able to rephrase it to avoid the double "extra columns", but I wanted to be clear that the metrics are coming from the extra columns.

**Comment by rdblue** [comment]
Thanks for the reviews, everyone! I've merged this.
